### PR TITLE
Open-ended ranges and rangeUntil operator proposal

### PR DESCRIPTION
**Pull request by ilya-g**
Discussion goes in [KEEP-314](https://github.com/Kotlin/KEEP/issues/314)
Please don't use this pull request for questions, only for proposals how to improve the text.

### CODE REVIEW

File: proposals/open-ended-ranges.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+```kotlin
+class IntRange : IntProgression(...), ClosedRange<Int>, OpenEndRange<Int> {
+    override val start: Int
+    override val endInclusive: Int
+    override val endExclusive: Int
+}
+```
+
+There's a subtlety in implementing `endExlcusive` property in such ranges: usually it returns `endInclusive + 1`, but
```
Comment (JakeWharton) [inline proposals/open-ended-ranges.md:158]:
```suggestion
There's a subtlety in implementing `endExclusive` property in such ranges: usually it returns `endInclusive + 1`, but 
```

Comment (ilya-g) [reply proposals/open-ended-ranges.md:158]: Thanks for spotting

File: proposals/open-ended-ranges.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+println(eqivalent)  // 1..1.9999999999999998
+```
+
+## Similar API review
+
+### Languages that distinguish end-inclusive and end-exclusive ranges
+
+- Swift: `...` end-inclusive range, `..<` end-exclusive, supports one-sided ranges 
+- Ruby: `..` end-inclusive range, `...` end-exclusive, supports one-sided ranges
+- Groovy: `..` end-inclusive range, `..<` end-exclusive, `<..` start-exclusive, `<..<` both bounds-exclusive range
```
Comment (elizarov) [inline proposals/open-ended-ranges.md:67]: A Rust is also worth noting here to add a variety to the syntax that different languages use (Rust is an outlier in using `..` for an end-exclusive range).

File: proposals/open-ended-ranges.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+```
+
+### Equality of open-ended ranges
+
+Similar to closed ranges, the `OpenEndRange` interface does not specify contract for `equals`/`hashCode` implementations,
+however its concrete implementations can do that. For example, an open-ended range of double values equal to another such range
+when bounds are respectively equal to each other, or to any empty range of doubles, when it is empty itself.
+
+Also, as a consequence of implementing both `OpenEndRange` and `ClosedRange` in concrete range implementations for standard discrete types,
+an open-ended range is equal to the closed range with the same `start` value and `endExlcusive` equal to `endInclusive + 1`:
```
Comment (hfhbd) [inline proposals/open-ended-ranges.md:198]:
```suggestion
an open-ended range is equal to the closed range with the same `start` value and `endExclusive` equal to `endInclusive + 1`:
```

Comment (ilya-g) [reply proposals/open-ended-ranges.md:198]: One was already fixed by Jake, now this is the last one! Thanks

File: proposals/open-ended-ranges.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+adjacent ranges neither have a point where they overlap, nor a point between them that is not contained in these ranges.  
+
+Even sometimes when the value is already discrete, for example, when it is expressed as a `Double` number, 
+and it is possible to represent a half-open range with a closed one by adjusting one of its bounds, 
+in practice it is not convenient to work with such ranges:
+
+```kotlin
+val equivalent = 1.0..2.0.nextDown() // contains the same values as 1.0..<2.0 range
+println(eqivalent)  // 1..1.9999999999999998
+```
```
[the remark this answers is not in the document]
Comment (ilya-g) [reply proposals/open-ended-ranges.md]: Yes, probably "discrete" is an unfortunate generalization term for the types like `Int`, `Long`, and `Char`. I've rewritten parts where it was used this way. 
`Double` is still a discrete number type, IMO. An example of non-discrete number would be a rational number represented by `BigInteger` numerator and denominator: no matter which two such numbers are given, we can always find another one between them.

**Pull request merged**